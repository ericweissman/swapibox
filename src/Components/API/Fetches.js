export const fetchData = async (url) => {
  const response = await fetch(url);
  if (response.ok) {
    return response.json();
  } else {
    throw Error(`Error fetching, ${response.status}`)
  }
}


