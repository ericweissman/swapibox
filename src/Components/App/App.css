main {
  display: -ms-grid;
  display: grid;
  -ms-grid-columns: 1fr 2fr;
      grid-template-columns: 1fr 2fr;
}
/*# sourceMappingURL=App.css.map */