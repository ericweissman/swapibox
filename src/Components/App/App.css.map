{
    "version": 3,
    "mappings": "AAAA,AAAA,IAAI,CAAC;EACH,OAAO,EAAE,IAAI;EACb,qBAAqB,EAAE,OAAO;CAC/B",
    "sources": [
        "App.scss"
    ],
    "names": [],
    "file": "App.css"
}