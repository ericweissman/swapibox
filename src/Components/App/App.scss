.App {
  background-image: url('https://i.stack.imgur.com/Mv83H.png');
  min-height: 700px;
}
main {
  display: grid;
  grid-template-columns: 1fr 2fr;
}

header {
  background-color: $menus;
}