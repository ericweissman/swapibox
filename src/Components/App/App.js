import React, { Component } from 'react';
import '../../Main.scss';
import MovieText from '../MovieText/MovieText';
import Controls from '../Controls/Controls';
import CardContainer from '../CardContainer/CardContainer';
import { fetchPeople, fetchData } from '../API/Fetches'


class App extends Component {
  constructor() {
    super();
    this.state = {
      films: {},
      people: [],
      planets: [],
      vehicles: [],
      active: 'people',
      errorStatus: ''
    }
  }

//PlANET FETCHES
fetchPlanetData = async () => {
  let allPlanets = [];
  const url = `https://swapi.co/api/planets/`;
  const planetData = await fetchData(url);
  allPlanets.push(...planetData.results)
  const planets = await this.fetchResidents(allPlanets)
  this.setState({ planets: planets })
}

fetchResidents = (planets) => {
  const unresolvedPromises = planets.map( async planet => {
    if (planet.residents.length > 0) {
     let residentNames = [];
     for (let i = 0; i < planet.residents.length; i++) {
       const residentData = await fetchData(planet.residents[i]);
       residentNames.push(residentData.name)
     }
      return ({
        name: planet.name,
        terrain: planet.terrain,
        population: planet.population,
        climate: planet.climate,
        residents: residentNames
      })
    } else {
      return ({
        name: planet.name,
        terrain: planet.terrain,
        population: planet.population,
        climate: planet.climate,
        residents: ['none']
      })
    }
  })
  return Promise.all(unresolvedPromises)
}


//PEOPLE FETCHES
fetchPeopleData = async () => {
  let allPeople = [];
  const url = `https://swapi.co/api/people/`

  try {
    const peopleData = await fetchData(url);
    allPeople.push(...peopleData.results)
    const withHome = await this.fetchHomeworlds(allPeople);
    const people = await this.fetchSpecies(withHome);
    this.setState({ people })
  } catch (error) {
    this.setState({errorStatus: error})
  }
}

fetchSpecies = (people) => {
  const unresolvedPromises = people.map( async (person) => {
    if (person.species.length > 0) {
      const speciesData =  await fetchData(person.species[0]);
      return ({
          name: person.name,
          homeworld: person.homeworld,
          population: person.population,
          species: speciesData.name
        })
    } else {
      return({
        name: person.name,
        homeworld: person.homeworld,
        population: person.population,
        species: 'unknown'
      })
    }
  })
  return Promise.all(unresolvedPromises)
}

fetchHomeworlds = (people) => {
  const unresolvedPromises = people.map( async (person) => {
      const homeworldData = await fetchData(person.homeworld);
      return ({ 
        ...person, 
        homeworld: homeworldData.name,
        population: homeworldData.population,
      })
  })
  return Promise.all(unresolvedPromises)
}

//MOVIE FETCH
fetchCrawl = async () => {
  let index =  Math.floor(Math.random() * 6 + 1)
  const url = 'https://swapi.co/api/films/';
  const filmData = await fetchData(url);
  const crawlData = await filmData.results[index];

  this.setState({
    films: {
      title: crawlData.title,
      crawl: crawlData.opening_crawl,
      year: crawlData.release_date
    }});
}


componentDidMount = () =>  {
  this.fetchPeopleData();
  this.fetchCrawl();
  this.fetchPlanetData();
}

  render() {
    const { films, planets, people, vehicles, active } = this.state;
    return (
      <div className="App">
        <header>
          <Controls />
        </header>
        <main>
          <MovieText films={films}/>
          <CardContainer 
            active={active}
            category={this.state}
            />
        </main>
      </div>
    );
  }
}

export default App;
