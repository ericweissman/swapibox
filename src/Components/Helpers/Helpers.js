import { fetchData } from "../API/Fetches";

//PEOPLE HELPERS
export const addHomeworlds = (people) => {
  const unresolvedPromises = people.map( async (person) =>{
    try {
      const homeworldData = await fetchData(person.homeworld);
      return ({
        ...person,
        homeworld: homeworldData.name,
        population: homeworldData.population
      })
    } catch (error){
      throw Error(error)
    }
  })
  return Promise.all(unresolvedPromises)
}

export const addSpecies = (people) => {
  const unresolvedPromises = people.map( async (person) => {
    if (person.species.length > 0) {
      try {
        const speciesData = await fetchData(person.species[0]);
        return ({
          name: person.name,
          homeworld: person.homeworld,
          population: person.population,
          species: speciesData.name
        })
      } catch (error) {
        throw Error(error)
      }

    } else {
      return ({
        name: person.name,
        homeworld: person.homeworld,
        population: person.population,
        species: 'unknown'
      })
    }
  })
  return Promise.all(unresolvedPromises)
}

export const chooseRandomFilm = async () => {
  const url = 'https://swapi.co/api/films/';
  try {
    const films = await fetchData(url);
    let index = await Math.floor(Math.random() * films.results.length);
    let crawl = await films.results[index];
    return crawl
  } catch (error) {
    return error
  }
}