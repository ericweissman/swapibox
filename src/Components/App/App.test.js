import React from 'react';
import ReactDOM from 'react-dom';
import App from './App';
import { shallow } from 'enzyme';
import * as Fetch from '../API/Fetches'
import * as Helpers from '../Helpers/Helpers'

describe('App', () => {
  let wrapper
  const defaultState = {
    films: {},
    people: [],
    planets: [],
    vehicles: [],
    favorites: [],
    active: '',
    errorStatus: ''
  }
  const mockCategory1 = 'people'
  const mockCategory2 = 'planets'
  const mockCategory3 = 'vehicles'
  

  beforeEach(() => {
    wrapper = shallow(<App />);
  })

  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should have the proper default state', () => {
    expect(wrapper.state()).toEqual(defaultState)
  });

  it('should correctly set the active category in state', () => {
    expect(wrapper.state('active')).toEqual('');
    wrapper.instance().updateActive(mockCategory1);
    expect(wrapper.state('active')).toEqual('people');
    wrapper.instance().updateActive(mockCategory2);
    expect(wrapper.state('active')).toEqual('planets');
  });

  it('should call the correct functions when populating data', () => {
    wrapper.instance().addPeople = jest.fn();
    wrapper.instance().addPlanets = jest.fn();
    wrapper.instance().addVehicles = jest.fn();

    wrapper.instance().populateData(mockCategory1);
    expect(wrapper.instance().addPeople).toHaveBeenCalledTimes(1);
    wrapper.instance().populateData(mockCategory2);
    expect(wrapper.instance().addPlanets).toHaveBeenCalledTimes(1);
    wrapper.instance().populateData(mockCategory3);
    expect(wrapper.instance().addVehicles).toHaveBeenCalledTimes(1);
  });

  it('should call the addCrawl method when the component mounts', () => {
    wrapper.instance().addCrawl = jest.fn();
    wrapper.instance().componentDidMount();
    expect(wrapper.instance().addCrawl).toBeCalled();
  });

  it('should add a movie to state when addCrawl is called', async () => {
    const film = {title: 'a', release_date: '1999', opening_crawl: 'texxxt'}
    const filmState = { title: 'a', year: '1999', crawl: 'texxxt' }
    Helpers.chooseRandomFilm = jest.fn(() => film)
    expect(wrapper.state()).toEqual(defaultState);
    await wrapper.instance().addCrawl();
    expect(wrapper.state('films')).toEqual(filmState)
  });

  it('should add people to the people array in state when addPeople is called', async () => {
    const people = [{name: 'Luke', homeworld: 'a', population: 'a', species: 'a', id: '1', favorite: false}]
    Helpers.addSpecies = jest.fn(() => people)
    jest.setTimeout(30000);
    wrapper.setState(defaultState)
    await wrapper.instance().addPeople();
    expect(wrapper.state('people')).toEqual(people)
  });

  it('should add planets to the planets array in state when addPlanets is called', async () => {
    const planets = [{name: 'a', terrain: 'a', population: 'a', climate: 'a', residents: ['a'], id: '2', favorite: false}]
    Helpers.fetchResidents = jest.fn(() => planets)
    jest.setTimeout(30000);
    wrapper.setState(defaultState);
    await wrapper.instance().addPlanets();
    expect(wrapper.state('planets')).toEqual(planets)
  })

  it('should add vehicles to the vehicles array in state when addVehicles is called', async () => {
    const vehicles = [{name: 'a', model: 'b', class: 'c', passengers: 'd', id: '2', favorite: false}];
    Helpers.cleanVehicles = jest.fn(() => vehicles);
    jest.setTimeout(30000);
    wrapper.setState(defaultState);
    await wrapper.instance().addVehicles();
    expect(wrapper.state('vehicles')).toEqual(vehicles)
  })

})


