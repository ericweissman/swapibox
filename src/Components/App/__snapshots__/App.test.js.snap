// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match snapshot 1`] = `
<div
  className="App"
>
  <header>
    <Controls />
  </header>
  <main>
    <MovieText
      films={Object {}}
    />
    <CardContainer
      active="planets"
      category={
        Object {
          "active": "planets",
          "films": Object {},
          "people": Array [],
          "planets": Array [],
          "vehicles": Array [],
        }
      }
    />
  </main>
</div>
`;
